crates/sui-json-rpc/src/error.rs

```diff
@@ -84,34 +85,133 @@ impl Error {
+    let error_object = ErrorObject::owned(-32603, err.to_string(), None::<String>);
+    RpcError::Call(CallError::Custom(error_object))
+}
+
+// One-to-one of the return type of the RPC method where applicable
+#[derive(Debug, Clone, Copy, AsRefStr, EnumString)]
+pub enum EntityType {
+    DelegatedStake,
+    SuiMoveNormalizedStruct,
+    SuiMoveNormalizedFunction,
+    Object,
+    Package,
+    Product,
+    SuiMoveNormalizedModule,
+    Checkpoint,
+}
+
+impl fmt::Display for EntityType {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.as_ref())
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum ClientError {
+    // Any kind of serialization or deserialization error
+    #[error("Serialization error on '{param}': {reason}")]
+    Serde { param: &'static str, reason: String },
+
+    #[error("Base64 error on '{param}': {reason}")]
+    Base64 { param: &'static str, reason: String },
+
+    #[error("BCS error on '{param}': {reason}")]
+    Bcs {
+        param: &'static str,
+        reason: bcs::Error,
+    },
+
+    #[error("FastCrypto error on '{param}': {reason}")]
+    FastCrypto {
+        param: &'static str,
+        reason: FastCryptoError,
+    },
 
-    #[error("request_type` must set to `None` or `WaitForLocalExecution` if effects is required in the response")]
+    #[error("Invalid combination of type and value: {0}")]
+    SerdeWithLayout(String),
+
+    #[error(transparent)]
+    Domain(#[from] DomainParseError),
+
+    #[error("Invalid '{param}': {reason}")]
+    InvalidParam { param: &'static str, reason: String },
+    // maybe InvalidParamMulti or something. param, value, reason
+    #[error("`request_type` must set to `None` or `WaitForLocalExecution` if effects is required in the response")]
     InvalidExecuteTransactionRequestType,
 
-    #[error("Unsupported protocol version requested. Min supported: {0}, max supported: {1}")]
-    ProtocolVersionUnsupported(u64, u64),
+    #[error("{entity} '{id}' not found")]
+    NotFound { entity: EntityType, id: String },
+
+    #[error("{entity} '{id}' not found in {in_entity} '{in_id}'")]
+    NotFoundIn {
+        entity: EntityType,
+        id: String,
+        in_entity: EntityType,
+        in_id: String,
+    },
 
-    #[error("Unable to serialize: {0}")]
-    CannotSerialize(#[from] bcs::Error),
+    #[error("{entity} '{id}' not found most likely due to pruning")]
+    Pruned { entity: EntityType, id: String },
 
+    // For error scenarios that don't fit cleanly into NotFound
     #[error("{0}")]
-    CannotParseSuiStructTag(String),
+    NotFoundCustom(String),
+
+    #[error("`{param}` exceeds limit of `{limit}`")]
+    LimitExceeded { param: &'static str, limit: usize },
+
+    #[error("`{param}` must be at least `{limit}`")]
+    LimitTooSmall { param: &'static str, limit: usize },
+
+    #[error("Invalid 'version': unsupported protocol version requested. Min supported: {0}, max supported: {1}")]
```

Comment:
is it invalid or unsupported? I could specify a valid protocol version that is out of date vs. not being a valid protocol version?

Reply:
Using the error message that previously existed might be the best here.

Reply:
I see, I was trying to keep the Invalid 'param': 'reason' but that might be too confusing